Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,177 @@
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel, bool generateId = true)
+        {
+            deviceModel.Created = DateTimeOffset.UtcNow;
+            deviceModel.Modified = deviceModel.Created;
+
+            if (generateId)
+            {
+                deviceModel.Id = Guid.NewGuid().ToString();
+            }
+
+            this.log.Debug("Creating a custom device model.", () => new { deviceModel });
+
+            // Note: using UpdateAsync because the service generates the ID
+            var result = await this.storage.UpdateAsync(
+                STORAGE_COLLECTION,
+                deviceModel.Id,
+                JsonConvert.SerializeObject(deviceModel),
+                null);
+
+            deviceModel.ETag = result.ETag;
+
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create or replace a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> UpsertAsync(DeviceModel deviceModel)
+        {
+            var id = deviceModel.Id;
+            var eTag = deviceModel.ETag;
+
+            try
+            {
+                var item = await this.GetAsync(id);
+
+                if (item.ETag == eTag)
+                {
+                    // Replace a custom device model
+                    deviceModel.Created = item.Created;
+                    deviceModel.Modified = DateTimeOffset.UtcNow;
+
+                    this.log.Debug("Modifying a custom device model via PUT.", () => new { deviceModel });
+
+                    var result = await this.storage.UpdateAsync(
+                        STORAGE_COLLECTION,
+                        id,
+                        JsonConvert.SerializeObject(deviceModel),
+                        eTag);
+
+                    // Return the new ETag provided by the storage
+                    deviceModel.ETag = result.ETag;
+                }
+                else
+                {
+                    this.log.Error("Invalid ETag.", () => new { CurrentETag = item.ETag, ETagProvided = eTag });
+                    throw new ConflictingResourceException("Invalid ETag. Device Model ETag is:'" + item.ETag + "'.");
+                }
+            }
+            catch (ResourceNotFoundException)
+            {
+                this.log.Info("Creating a new device model via PUT.", () => new { deviceModel });
+
+                var result = await this.InsertAsync(deviceModel, false);
+                deviceModel.ETag = result.ETag;
+            }
+            catch (Exception exception)
+            {
+                this.log.Error("Something went wrong while upserting the device model.", () => new { exception });
```

Comment:
in the previous review the exception was rethrown, now it's swallowed. Any reason for the change?  Looks like a bug, there might not be a model to return at line 166

Reply:
nit: include the device model in the log, like line 156

Reply:
Done.